### DISCUSSION TITLE
Allow for mutations from Python

### DISCUSSION THREAD

**Discussion by tschiex** [BradyAJohnston/MolecularNodes#127] [Ideas]
**Is your feature request related to a problem? Please describe.**
As a protein designer, I often need to mutate amino acids of existing proteins. It would be nice to be able to do this kind of mutations from Blender in Python using molecular nodes. This probably requires to use a discrete or constinuous rotamer library such as Dunback's of Basilisk. 

**Describe the solution you'd like**
The ability to set the amino acid at a given ResId and then control its geometry through internal coordinates or from successive choices in a rotamer library.

**Describe alternatives you've considered**
This is feasible in PyMol using the Mutagenesis wizard.

Using PyRosetta here could be handy (but it's a heavy non-open-source dependency).

**Comment by BradyAJohnston** [comment]
Internally I use [Biotite](https://www.biotite-python.org) which would potentially enable this kind of thing. 

As it currently stands, the workflow takes the protein coordinate information and converts is into Blender objects via Biotite or MDAnalysis, then once it is inside of Blender it is all based on Geometry Nodes.

To support this there would have be a back and forth between Blender models and Biotite AtomArrays. This shouldn't actually be too difficult in principle, but there isn't any foundation established at the moment for doing so.

I have stayed away from a lot of specific functionality that other programs like PyMol and ChimeraX enable, trying to focus this being purely a visualisation package. Adding more and more features like this will increase code base and development time. For something like mutating residues though I think this could be part of a broader protein manipulation workflow potentially. Open to suggestions, but also currently pretty limited on dev time so I wouldn't say it's planned any time soon.

**Comment by tschiex** [comment]
As a developer, I absolutely understand your point of view. Don't get bloated too quickly! 
I couldn't install MolecularNodes 0.13 on my Ubuntu machine (I wanted to look at all the functionalities that are here but not in 2.0) but 2.0 installed itself smoothly and I may try to play with it and PyRosetta together, just for fun, if only I can find some free time...

**Comment by BradyAJohnston** [comment]
I have no experience with PyRosetta, so if you end up figuring something out that could work, then I'd love to know about it! :)

**Comment by YaoYinYing** [comment]
I really like this idea but it will not be an easy one.
The question can be divided into the following parts:
- how to install PyRosetta as a dependency of MN.
- how to manipulate (or let's say, update) the coordinates of imported molecule (in molecule object class?) in Blender.
- how to parse the mutation input string into an operator.
- how we can provide the access(a pop input box? a functional tool in side bar?)  of molecule mutagenesis to user.

 Unfortunately I'm still an intro of Rosetta and not familiar with PyRosetta. The installation part is implemented as in [this commit](https://github.com/YaoYinYing/MolecularNodes/commit/c22ffc71d21a5186008980aee8b2395ea07846dc). It works in most location of the planet but needs some cleanup.

Should we look into the PyMOL-PyRoseeta server script for hints?

**Comment by kimjoyc** [comment]
@BradyAJohnston  Is there a way to provide dynamism for various structures you want to load even after the load without exiting Blender everytime? 

The reason I ask is that I noticed using the ea_fec.pdb and ea_fec.dcd, I can only load that selection on that structure one at a time. 

For instance, in custom_selections, is there a way for me to visualize multiple selections simultaneously?

**Reply by BradyAJohnston** [reply]
I am unsure what you are actually asking here. You can create multiple custom selections at once on import. Creating new selections once imported pretty much has to be done via geometry nodes.

**Comment by tschiex** [comment]
Dear Brady,

I’m in the process of submitting a paper that uses figures made with "Molecular Nodes ».
I explored https://bradyajohnston.github.io/MolecularNodes/ but didn’t find any advice:

What would be the best way to thank/cite you?

Thomas

**Reply by BradyAJohnston** [reply]
Thanks for asking! There still isn't a paper published on it, but each version gets a Zenodo DOI generated (for example most recent version: https://zenodo.org/records/14241983) which you can use for citing.

**Comment by tschiex** [comment]
Hello Brady,

If you are interested in protein design, you can have a look to our paper (with MolecularNodes images and references to the add-on in the acknowledgements) here: 

	https://rdcu.be/ebvNU <https://rdcu.be/ebvNU>

Thanks again for you work !

Thomas

> Thanks for asking! There still isn't a paper published on it, but each version gets a Zenodo DOI generated (for example most recent version: https://zenodo.org/records/14241983) which you can use for citing.

**Reply by BradyAJohnston** [reply]
Really beautiful figures, thanks for the sharing the link to the paper!

**Reply by BradyAJohnston** [reply]
Any chance I could get access to the PDF? I can see it through the online but would love to save in Zotero

**Comment by tschiex** [comment]
Hello Brady,

I would love to be able to send the PDF to you, but the Nature publishing group is happy to exploit its strong  visibility to « constrain » scientific authors.
The authors are required to respect a 6 months embargo on the PDF availabilit. We are instead allowed to send this « link » where you can easily browse the paper, but not download a PDF.

As a computer person, I would be confident a very decent PDF could be extracted from this (even better using AI) but someone has yet to build the extraction tool.

Kind regards,

Thomas
